File: spark/panalgo/src/main/scala/panalgo/Panalgo.scala
package panalgo

import org.apache.spark.sql.functions._
import org.apache.spark.sql.types.{DataTypes, StructField, StructType}
import org.apache.spark.sql.{DataFrame, Row, SparkSession}

import java.sql.Date
import java.time.Period
import scala.io.Source

object Panalgo {
  val spark = SparkSession.builder.appName("Panalgo").master("local[*]").getOrCreate

  val PATIENT_DATE_FIELD = "CLM_FROM_DT"
  val PATIENT_ID_FIELD = "DESYNPUF_ID"
  val PRESCRIPTION_DATE_FIELD = "SRVC_DT"
  val NDC_ID_FIELD = "PROD_SRVC_ID"
  val DIABETES_INDEX_DATE = "DIABETES_INDEX_DATE"

  val inpatientFile = "data/raw/inpatient/inpatient.txt"
  val outpatientFile = "data/raw/outpatient/outpatient.txt"
  val carrierFile = "data/raw/carrier/carrier.txt"
  val beneficiaryFile = "data/raw/beneficiary/beneficiary.txt"
  val prescriptionFile = "data/raw/prescription/prescription.txt"
  val lovastatinFile = "data/lookup/lovastatin.txt"

  val lovastatin = Source.fromFile(lovastatinFile).getLines.toList

  // diagnosis column names
  val inpatientDiagnosisColNames = List("ADMTNG_ICD9_DGNS_CD", "ICD9_DGNS_CD_1", "ICD9_DGNS_CD_2", "ICD9_DGNS_CD_3", "ICD9_DGNS_CD_4", "ICD9_DGNS_CD_5", "ICD9_DGNS_CD_6", "ICD9_DGNS_CD_7", "ICD9_DGNS_CD_8", "ICD9_DGNS_CD_9", "ICD9_DGNS_CD_10")
  val outpatientDiagnosisColNames = List("ADMTNG_ICD9_DGNS_CD", "ICD9_DGNS_CD_1", "ICD9_DGNS_CD_2", "ICD9_DGNS_CD_3", "ICD9_DGNS_CD_4", "ICD9_DGNS_CD_5", "ICD9_DGNS_CD_6", "ICD9_DGNS_CD_7", "ICD9_DGNS_CD_8", "ICD9_DGNS_CD_9", "ICD9_DGNS_CD_10")
  val carrierDiagnosisColNames = List("ICD9_DGNS_CD_1", "ICD9_DGNS_CD_2", "ICD9_DGNS_CD_3", "ICD9_DGNS_CD_4", "ICD9_DGNS_CD_5", "ICD9_DGNS_CD_6", "ICD9_DGNS_CD_7", "ICD9_DGNS_CD_8", "LINE_ICD9_DGNS_CD_1", "LINE_ICD9_DGNS_CD_2", "LINE_ICD9_DGNS_CD_3", "LINE_ICD9_DGNS_CD_4", "LINE_ICD9_DGNS_CD_5", "LINE_ICD9_DGNS_CD_6", "LINE_ICD9_DGNS_CD_7", "LINE_ICD9_DGNS_CD_8", "LINE_ICD9_DGNS_CD_9", "LINE_ICD9_DGNS_CD_10", "LINE_ICD9_DGNS_CD_11", "LINE_ICD9_DGNS_CD_11", "LINE_ICD9_DGNS_CD_13")

  val yyyyddmmToAcceptableDateUDF = udf((s: String) =>
    if (s != null)
      s.substring(0, 4) + "-" + s.substring(4, 6) + "-" + s.substring(6, 8)
    else
      null)

  val age = udf((date: Date, dob: Date) => Period.between(dob.toLocalDate, date.toLocalDate).getYears)

  def diabeticPatients() = {
    val inpatient = spark.read.format("csv").option("header", "true").load(inpatientFile)
    val outpatient = spark.read.format("csv").option("header", "true").load(outpatientFile)
    val carrier = spark.read.format("csv").option("header", "true").load(carrierFile)
    val schema = StructType(StructField(PATIENT_ID_FIELD, DataTypes.StringType, true) :: Nil)
    var diabeticPatients = spark.createDataFrame(spark.sparkContext.emptyRDD[Row], schema)
    for (name <- inpatientDiagnosisColNames) {
      diabeticPatients = diabeticPatients.union(
        inpatient
          .filter(col(PATIENT_DATE_FIELD).startsWith("2009") && col(name).startsWith("250"))
          .select(col(PATIENT_ID_FIELD)))
    }
    for (name <- outpatientDiagnosisColNames) {
      diabeticPatients = diabeticPatients.union(
        outpatient
          .filter(col(PATIENT_DATE_FIELD).startsWith("2009") && col(name).startsWith("250"))
          .select(col(PATIENT_ID_FIELD)))
    }
    for (name <- carrierDiagnosisColNames) {
      diabeticPatients = diabeticPatients.union(
        carrier
          .filter(col(PATIENT_DATE_FIELD).startsWith("2009") && col(name).startsWith("250"))
          .select(col(PATIENT_ID_FIELD)))
    }
    diabeticPatients.dropDuplicates(PATIENT_ID_FIELD)
  }

  def dateIndexedDiabeticPatients(): DataFrame = {
    val inpatient = spark.read.format("csv").option("header", "true").load(inpatientFile)
    val outpatient = spark.read.format("csv").option("header", "true").load(outpatientFile)
    val carrier = spark.read.format("csv").option("header", "true").load(carrierFile)
    val schema = StructType(
      StructField(PATIENT_ID_FIELD, DataTypes.StringType, true) ::
        StructField(PATIENT_DATE_FIELD, DataTypes.StringType, true) ::
        StructField(DIABETES_INDEX_DATE, DataTypes.DateType, true) :: Nil)
    var datedDiabeticPatients = spark.createDataFrame(spark.sparkContext.emptyRDD[Row], schema)
    for (name <- inpatientDiagnosisColNames) {
      datedDiabeticPatients = datedDiabeticPatients.union(
        inpatient
          .filter(col(PATIENT_DATE_FIELD).startsWith("2009") && col(name).startsWith("250"))
          .select(col(PATIENT_ID_FIELD), col(PATIENT_DATE_FIELD), to_date(yyyyddmmToAcceptableDateUDF(col(PATIENT_DATE_FIELD)), "yyyy-MM-dd")))
    }
    for (name <- outpatientDiagnosisColNames) {
      datedDiabeticPatients = datedDiabeticPatients.union(
        outpatient
          .filter(col(PATIENT_DATE_FIELD).startsWith("2009") && col(name).startsWith("250"))
          .select(col(PATIENT_ID_FIELD), col(PATIENT_DATE_FIELD), to_date(yyyyddmmToAcceptableDateUDF(col(PATIENT_DATE_FIELD)), "yyyy-MM-dd")))
    }
    for (name <- carrierDiagnosisColNames) {
      datedDiabeticPatients = datedDiabeticPatients.union(
        carrier
          .filter(col(PATIENT_DATE_FIELD).startsWith("2009") && col(name).startsWith("250"))
          .select(col(PATIENT_ID_FIELD), col(PATIENT_DATE_FIELD), to_date(yyyyddmmToAcceptableDateUDF(col(PATIENT_DATE_FIELD)), "yyyy-MM-dd")))
    }
    datedDiabeticPatients = datedDiabeticPatients.groupBy(col(PATIENT_ID_FIELD))
      .agg(min(DIABETES_INDEX_DATE).as(DIABETES_INDEX_DATE))
      .drop(PATIENT_DATE_FIELD)
    datedDiabeticPatients
  }

  def lovastatinPrescriptions(): DataFrame = {
    val prescription = spark.read.format("csv").option("header", "true").load(prescriptionFile)
    prescription
      .select(col(PATIENT_ID_FIELD), to_date(yyyyddmmToAcceptableDateUDF(col(PRESCRIPTION_DATE_FIELD)), "yyyy-MM-dd").as(PRESCRIPTION_DATE_FIELD), col(NDC_ID_FIELD))
      .filter(col(NDC_ID_FIELD).isin(lovastatin: _*))
  }

  def dateIndexedDiabeticPatientsWithLovastatinPrescriptions(): DataFrame = {
    val patients = dateIndexedDiabeticPatients()
    val prescriptions = lovastatinPrescriptions()
    patients
      .withColumn("DT", add_months(col(DIABETES_INDEX_DATE), 12))
      .join(prescriptions, patients.col(PATIENT_ID_FIELD).equalTo(prescriptions.col(PATIENT_ID_FIELD)))
      .drop(prescriptions.col(PATIENT_ID_FIELD))
      .filter(col(PRESCRIPTION_DATE_FIELD).gt(col(DIABETES_INDEX_DATE)))
      .filter(col(PRESCRIPTION_DATE_FIELD).lt(col("DT")))
      .drop(col("DT"))
      .drop(col(PRESCRIPTION_DATE_FIELD))
      .drop(col(NDC_ID_FIELD))
  }

  def retiredDateIndexedDiabeticPatientsWithLovastatinPrescriptions(): DataFrame = {
    val patients = dateIndexedDiabeticPatientsWithLovastatinPrescriptions()
    val beneficiary = spark.read.format("csv").option("header", "true").load(beneficiaryFile)
    val retired = patients
      .join(beneficiary, patients.col(PATIENT_ID_FIELD).equalTo(beneficiary.col(PATIENT_ID_FIELD)))
      .drop(beneficiary.col("BENE_SEX_IDENT_CD"))
      .drop(beneficiary.col("BENE_RACE_CD"))
      .drop(beneficiary.col("BENE_ESRD_IND"))
      .drop(beneficiary.col("BENE_DEATH_DT"))
      .select(patients.col(PATIENT_ID_FIELD),
        patients.col(DIABETES_INDEX_DATE),
        to_date(yyyyddmmToAcceptableDateUDF(col("BENE_BIRTH_DT")), "yyyy-MM-dd").as("BENE_BIRTH_DT"),
      )
      .withColumn("AGE", age(col(DIABETES_INDEX_DATE), col("BENE_BIRTH_DT")))
    retired.filter(col("AGE").geq(65))
      .select(patients.col(PATIENT_ID_FIELD))
      .dropDuplicates(PATIENT_ID_FIELD)
  }

  def main(args: Array[String]): Unit = {
    // Node 1: Diabetes Patients -- 1740
    val node1 = diabeticPatients().count()

    // Node 2: Lovastatin -- 142
    val node2 = dateIndexedDiabeticPatientsWithLovastatinPrescriptions()
      .drop(col(DIABETES_INDEX_DATE))
      .dropDuplicates(PATIENT_ID_FIELD)
      .count()

    // Node 3: Patients Age > 65 -- 119
    val node3 = retiredDateIndexedDiabeticPatientsWithLovastatinPrescriptions().count()

    println(s"Section              Count")
    println(s"Node 1 (Diabetes):   $node1")
    println(s"Node 2 (Lovastatin): $node2")
    println(s"Node 3 (Age > 65):   $node3")
  }
}

File: python/panalgo.py
#!/usr/bin/env python3

import csv
import datetime
from itertools import groupby

PATIENT_DATE_FIELD = "CLM_FROM_DT"
PATIENT_ID_FIELD = "DESYNPUF_ID"
PRESCRIPTION_DATE_FIELD = "SRVC_DT"
NDC_ID_FIELD = "PROD_SRVC_ID"
DIABETES_INDEX_DATE = "DIABETES_INDEX_DATE"

inpatientDiagnosisColNames = ["ADMTNG_ICD9_DGNS_CD", "ICD9_DGNS_CD_1", "ICD9_DGNS_CD_2", "ICD9_DGNS_CD_3",
                              "ICD9_DGNS_CD_4", "ICD9_DGNS_CD_5", "ICD9_DGNS_CD_6", "ICD9_DGNS_CD_7",
                              "ICD9_DGNS_CD_8", "ICD9_DGNS_CD_9", "ICD9_DGNS_CD_10"]
outpatientDiagnosisColNames = ["ADMTNG_ICD9_DGNS_CD", "ICD9_DGNS_CD_1", "ICD9_DGNS_CD_2", "ICD9_DGNS_CD_3",
                               "ICD9_DGNS_CD_4", "ICD9_DGNS_CD_5", "ICD9_DGNS_CD_6", "ICD9_DGNS_CD_7",
                               "ICD9_DGNS_CD_8", "ICD9_DGNS_CD_9", "ICD9_DGNS_CD_10"]
carrierDiagnosisColNames = ["ICD9_DGNS_CD_1", "ICD9_DGNS_CD_2", "ICD9_DGNS_CD_3", "ICD9_DGNS_CD_4",
                            "ICD9_DGNS_CD_5", "ICD9_DGNS_CD_6", "ICD9_DGNS_CD_7", "ICD9_DGNS_CD_8",
                            "LINE_ICD9_DGNS_CD_1", "LINE_ICD9_DGNS_CD_2", "LINE_ICD9_DGNS_CD_3",
                            "LINE_ICD9_DGNS_CD_4", "LINE_ICD9_DGNS_CD_5", "LINE_ICD9_DGNS_CD_6",
                            "LINE_ICD9_DGNS_CD_7", "LINE_ICD9_DGNS_CD_8", "LINE_ICD9_DGNS_CD_9",
                            "LINE_ICD9_DGNS_CD_10", "LINE_ICD9_DGNS_CD_11", "LINE_ICD9_DGNS_CD_11",
                            "LINE_ICD9_DGNS_CD_13"]

inpatientFile = "../spark/data/raw/inpatient/inpatient.txt"
outpatientFile = "../spark/data/raw/outpatient/outpatient.txt"
carrierFile = "../spark/data/raw/carrier/carrier.txt"
beneficiaryFile = "../spark/data/raw/beneficiary/beneficiary.txt"
prescriptionFile = "../spark/data/raw/prescription/prescription.txt"
lovastatinFile = "../spark/data/lookup/lovastatin.txt"

class Panalgo:

    def read_csv(self, file):
        with open(file, newline='') as csvfile:
            reader = csv.DictReader(csvfile)
            return list(reader)

    def __init__(self):
        self.inpatient = self.read_csv(inpatientFile)
        self.outpatient = self.read_csv(outpatientFile)
        self.carrier = self.read_csv(carrierFile)
        self.beneficiary = self.read_csv(beneficiaryFile)
        self.prescription = self.read_csv(prescriptionFile)
        with open(lovastatinFile) as f:
            self.lovastatin = set(f.read().splitlines())

    def diabeticPatients(self):
        diabetic_patients = []
        for col_name in inpatientDiagnosisColNames:
            for patient in self.inpatient:
                if patient[PATIENT_DATE_FIELD].startswith('2009') and patient[col_name].startswith('250'):
                    diabetic_patient = {
                        PATIENT_ID_FIELD: patient[PATIENT_ID_FIELD],
                        DIABETES_INDEX_DATE: patient[PATIENT_DATE_FIELD],
                    }
                    diabetic_patients.append(diabetic_patient)
        for col_name in outpatientDiagnosisColNames:
            for patient in self.outpatient:
                if patient[PATIENT_DATE_FIELD].startswith('2009') and patient[col_name].startswith('250'):
                    diabetic_patient = {
                        PATIENT_ID_FIELD: patient[PATIENT_ID_FIELD],
                        DIABETES_INDEX_DATE: patient[PATIENT_DATE_FIELD],
                    }
                    diabetic_patients.append(diabetic_patient)
        for col_name in carrierDiagnosisColNames:
            for patient in self.carrier:
                if patient[PATIENT_DATE_FIELD].startswith('2009') and patient[col_name].startswith('250'):
                    diabetic_patient = {
                        PATIENT_ID_FIELD: patient[PATIENT_ID_FIELD],
                        DIABETES_INDEX_DATE: patient[PATIENT_DATE_FIELD],
                    }
                    diabetic_patients.append(diabetic_patient)
        return diabetic_patients

    def dateIndexediabeticPatients(self):
        diabetic_patients = [(dp[PATIENT_ID_FIELD], dp[DIABETES_INDEX_DATE]) for dp in self.diabeticPatients()]
        diabetic_patients.sort(key=lambda item: (item[0], item[1]))
        with_diabetes_index_date = []
        grouped = groupby(diabetic_patients, key=lambda item: item[0])
        for patient_id, group in grouped:
            grp = list(group)
            diabetic_patient = {
                PATIENT_ID_FIELD: patient_id,
                DIABETES_INDEX_DATE: grp[0][1],
            }
            with_diabetes_index_date.append(diabetic_patient)
        return with_diabetes_index_date

    def lovastatinPrescriptions(self):
        return [(p[PATIENT_ID_FIELD], p[PRESCRIPTION_DATE_FIELD]) for p in
                self.prescription if p[NDC_ID_FIELD] in self.lovastatin]

    def DiabeticPatientsWithLovastatinPrescriptions(self):
        lovastatin_patients = set()
        diabetic_patients = dict(
            [(dp[PATIENT_ID_FIELD], dp[DIABETES_INDEX_DATE]) for dp in self.dateIndexediabeticPatients()])
        prescriptions = self.lovastatinPrescriptions()
        for patient_id, prescription_date in prescriptions:
            if patient_id in diabetic_patients:
                diabetes_index_date = diabetic_patients[patient_id]
                diabetes_index_date_plus_year = str(int(diabetes_index_date[0:4]) + 1) + diabetes_index_date[4:]
                if prescription_date > diabetes_index_date and prescription_date < diabetes_index_date_plus_year:
                    lovastatin_patients.add(patient_id)
        return dict([(patient_id, diabetic_patients[patient_id]) for patient_id in lovastatin_patients])

    def retiredDiabeticPatientsWithLovastatinPrescriptions(self):
        retired_patients = {}
        lovastatin_patients = self.DiabeticPatientsWithLovastatinPrescriptions()
        beneficiaries = dict([(b[PATIENT_ID_FIELD], b['BENE_BIRTH_DT']) for b in self.beneficiary])
        for pid, date in lovastatin_patients.items():
            if pid in beneficiaries:
                diabetes_index_date = datetime.datetime.strptime(date, "%Y%m%d")
                birth_date = datetime.datetime.strptime(beneficiaries.get(pid), "%Y%m%d")
                delta = ((diabetes_index_date - birth_date).days) / 365
                if delta >= 65:
                    retired_patients[pid] = date
        return retired_patients

    def run(self):
        diabetic_patients = self.diabeticPatients()
        node1 = len(set([dp[PATIENT_ID_FIELD] for dp in diabetic_patients]))
        lovastatin_patients = self.DiabeticPatientsWithLovastatinPrescriptions()
        node2 = len(lovastatin_patients.keys())
        retired_patients = self.retiredDiabeticPatientsWithLovastatinPrescriptions()
        node3 = len(set(retired_patients.keys()))
        print(f"Section              Count")
        print(f"Node 1 (Diabetes):   {node1}")
        print(f"Node 2 (Lovastatin): {node2}")
        print(f"Node 3 (Age > 65):   {node3}")

if __name__ == "__main__":
    Panalgo().run()
